solidity/contracts/converter/types/standard-pool/StandardPoolConverter.sol

```diff
@@ -0,0 +1,1362 @@
+            _beneficiary.transfer(amount);
+        else
+            safeTransfer(_targetToken, _beneficiary, amount);
+
+        // dispatch the conversion event
+        dispatchConversionEvent(_sourceToken, _targetToken, _trader, _amount, amount, fee);
+
+        // dispatch rate updates
+        dispatchTokenRateUpdateEvents(_sourceToken, _targetToken, actualSourceBalance, targetBalance - amount);
+
+        return amount;
+    }
+
+    /**
+      * @dev returns the recent average rate of 1 `_token` in the other reserve token units
+      *
+      * @param _token   token to get the rate for
+      * @return recent average rate between the reserves (numerator)
+      * @return recent average rate between the reserves (denominator)
+    */
+    function recentAverageRate(IERC20Token _token) external view returns (uint256, uint256) {
+        // get the recent average rate of reserve 0
+        uint256 rate = calcRecentAverageRate(averageRateInfo);
+        if (_token == __reserveTokens[0]) {
+            return ((rate >> 112) & MAX_UINT112, rate & MAX_UINT112);
+        }
+
+        return (rate & MAX_UINT112, (rate >> 112) & MAX_UINT112);
+    }
+
+    /**
+      * @dev updates the recent average rate if needed
+    */
+    function updateRecentAverageRate() internal {
+        uint256 averageRateInfo1 = averageRateInfo;
+        uint256 averageRateInfo2 = calcRecentAverageRate(averageRateInfo1);
+        if (averageRateInfo1 != averageRateInfo2) {
+            averageRateInfo = averageRateInfo2;
+        }
+    }
+
+    /**
+      * @dev returns the recent average rate of 1 reserve token 0 in reserve token 1 units
+      *
+      * @param _prevAverageRate a local copy of the previous average rate state-variable
+      * @return recent average rate between the reserves
+    */
+    function calcRecentAverageRate(uint256 _prevAverageRate) internal view returns (uint256) {
+        // get the previous average rate and its update-time
+        uint256 prevAverageRateT = _prevAverageRate >> 224;
+        uint256 prevAverageRateN = (_prevAverageRate >> 112) & MAX_UINT112;
+        uint256 prevAverageRateD = _prevAverageRate & MAX_UINT112;
+
+        // get the elapsed time since the previous average rate was calculated
+        uint256 currentTime = time();
+        uint256 timeElapsed = currentTime - prevAverageRateT;
+
+        // if the previous average rate was calculated in the current block, the average rate remains unchanged
+        if (timeElapsed == 0) {
+            return _prevAverageRate;
+        }
+
+        // get the current rate between the reserves
+        uint256 currentRate = __reserveBalances;
+        uint256 currentRateN = currentRate >> 128;
+        uint256 currentRateD = currentRate & MAX_UINT128;
+
+        // if the previous average rate was calculated a while ago or never, the average rate is equal to the current rate
+        if (timeElapsed >= AVERAGE_RATE_PERIOD || prevAverageRateT == 0) {
```

Comment:
can we save some gas by creating a local var for AVERAGE_RATE_PERIOD? can't remember if constants are in the binary or in the storage

Reply:
Constants are replaced prior to compilation, so they are obviously not in storage.
In the past, solc used to replace the constant expression as is, so if that expression had any computations within it (for example, `2 ** 128 - 1`, then they would "find their way" into the actual code (instead of having the actual constant value everywhere).
Not sure how it is now, but if any gas issue with constants, then it is this issue and not a (much worse) storage issue.
And in this case, since the constant expression at hand (`AVERAGE_RATE_PERIOD`) does not consist of any internal computations, even this potential issue is irrelevant.

Reply:
yeah I was referring to SLOADs

Reply:
No one would have used constants if that had been the case.